lib/ansible/modules/monitoring/grafana_datasource.py

```diff
@@ -0,0 +1,469 @@
+            r, info = fetch_url(module, '{}/api/datasources/{}'.format(data['grafana_url'], ds['id']), data=json.dumps(payload), headers=headers, method='PUT')
+            if info['status'] == 200:
+                res = json.loads(r.read())
+                result['name'] = data['name']
+                result['id'] = ds['id']
+                result['before'] = ds
+                result['after'] = payload
+                result['msg'] = "Datasource {} updated {}".format(data['name'], res['message'])
+                result['changed'] = True
+            else:
+                raise GrafanaAPIException('Unable to update the datasource id {} : {}'.format(ds['id'], info))
+    else:
+        # create
+        r, info = fetch_url(module, '{}/api/datasources'.format(data['grafana_url']), data=json.dumps(payload), headers=headers, method='POST')
+        if info['status'] == 200:
+            res = json.loads(r.read())
+            result['msg'] = "Datasource {} created : {}".format(data['name'], res['message'])
+            result['changed'] = True
+            result['name'] = data['name']
+            result['id'] = res['id']
+        else:
+            raise GrafanaAPIException('Unable to create the new datasource {} : {} - {}.'.format(data['name'], info['status'], info))
+
+    return result
+
+
+def grafana_delete_datasource(module, data):
+
+    # define http headers
+    headers = {'content-type': 'application/json'}
+    if 'grafana_api_key' in data and data['grafana_api_key']:
+        headers['Authorization'] = "Bearer {}".format(data['grafana_api_key'])
+    else:
+        auth = base64.encodestring('%s:%s' % (data['grafana_user'], data['grafana_password'])).replace('\n', '')
+        headers['Authorization'] = 'Basic %s' % auth
+        grafana_switch_organisation(module, data['grafana_url'], data['org_id'], headers)
+
+    # test if datasource already exists
+    datasource_exists, ds = grafana_datasource_exists(module, data['grafana_url'], data['name'], headers=headers)
+
+    result = {}
+    if datasource_exists is True:
+        # delete
+        r, info = fetch_url(module, '{}/api/datasources/name/{}'.format(data['grafana_url'], data['name']), headers=headers, method='DELETE')
+        if info['status'] == 200:
+            res = json.loads(r.read())
+            result['msg'] = "Datasource {} deleted : {}".format(data['name'], res['message'])
+            result['changed'] = True
+            result['name'] = data['name']
+            result['id'] = 0
+        else:
+            raise GrafanaAPIException('Unable to update the datasource id {} : {}'.format(ds['id'], info))
+    else:
+        # datasource does not exists : do nothing
+        result = {'msg': "Datasource {} does not exists".format(data['name']),
+                  'changed': False,
+                  'id': 0,
+                  'name': data['name']}
+
+    return result
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
```

Comment:
Please update as per
https://github.com/ansible/ansible/pull/31930/files

```
required_together=[['grafana_user', 'grafana_password', 'org_id']],
mutually_exclusive=[['grafana_user', 'grafana_api_key']],
```

Reply:
I've also added `required_if` for some parameters..